---
title: "Using the same procedures DESeq2 used to compare its analysis to other models, my Bayesian method is compare to Deseq2."
output: html_notebook
---

# Repeat DESeq2 paper figure 6

Recreate the comparison graph between different analyses given in the DESeq 2 paper!

```{r deseq2SimulatedDataComparison}
load("../../data/experimental/meanDispPairsPickrellEtAl.RData")

library("BiocParallel")
library("rstan")
library("tidyverse")
library("Biobase")
library("DESeq")
library("DESeq2")
library("edgeR")
library("limma")
library("samr")
library("DSS")
library("EBSeq")
source("runDEalgorithmsScript.R")

register(MulticoreParam(workers=4)) 

# A function that creates transcript counts for n genes. m/2 is an integer representing the number of replicates per condition. x is a model matrix (with nCol == No of Conditions and nRow / nCol == No of reps per condition). beta is a vector of length n, each element represent the log2 fold change between the conditions (0 = no fold change). meanDispPairs is a list of the means and dispersions associated with transcripts of all genes in the human genome deduced from the Pickrell et al 2010 dataset.

makeSimData <- function(n, m, x, beta, meanDispPairs, sf=rep(1,m)) {
  idx <- sample(nrow(meanDispPairs), n, replace=TRUE)
  mu0 <- meanDispPairs[idx,1]
  disp <- meanDispPairs[idx,2]
  betafull <- cbind(log2(mu0), beta)
  mu <- 2^(betafull %*% t(x))
  muMat <- matrix(rep(mu, times=m) * rep(sf, each=n), ncol=m)
  list(mat = matrix(rnbinom(n*m, mu=muMat, size=1/disp), ncol=m),
       disp = disp,
       mu0 = mu0)
}



algos <- list("Bayesian"=runBayes,"DESeq"=runDESeq,"DESeq2"=runDESeq2,"edgeR"=runEdgeR)#,"edgeR-robust"=runEdgeRRobust,
              #"DSS"=runDSS,"voom"=runVoom,"SAMseq"=runSAMseq,"EBSeq"=runEBSeq)
namesAlgos <- names(algos)

n <- 1000
effSizeLevels <- log2(c(2,3,4)) # create multiple datasets with different magnitudes of log-fold changes between conditions
mLevels <- c(6,8,10,20) # create multiple datasets with different numbers of replicates within conditions
nreps <- 3 # create six different datasets for each log fold change magnitude and each replicate number (to compare consistancy)

effSizes <- rep(rep(effSizeLevels, each=nreps), times=length(mLevels))
ms <- rep(mLevels, each=nreps * length(effSizeLevels))

resList <- bplapply(seq_along(ms), function(i) {
  set.seed(i)
  m <- ms[i]
  es <- effSizes[i]
  condition <- factor(rep(c("A","B"), each = m/2))
  x <- model.matrix(~ condition)
  beta <- c(rep(0, n * 8/10), sample(c(-es,es), n * 2/10, TRUE))
  mat <- makeSimData(n,m,x,beta,meanDispPairs)$mat
  e <- ExpressionSet(mat, AnnotatedDataFrame(data.frame(condition)))
  resTest <- lapply(algos, function(f) f(e))
  nonzero <- rowSums(exprs(e)) > 0
  sensidx <- abs(beta) > 0 & nonzero
  sens <- sapply(resTest, function(z) mean((z$padj < .1)[sensidx]))
  rmf <- cut(rowMeans(mat), c(0, 20, 100, 300, Inf), include.lowest=TRUE)
  levels(rmf) <- paste0("sens",c("0to20","20to100","100to300","more300"))
  sensStratified <- t(sapply(resTest, function(z) tapply( (z$padj < .1)[sensidx], rmf[sensidx], mean)))
  oneminusspecpvals <- sapply(resTest, function(z) mean((z$pvals < .01)[beta == 0 & nonzero], na.rm=TRUE))
  oneminusspecpadj <- sapply(resTest, function(z) mean((z$padj < .1)[beta == 0 & nonzero], na.rm=TRUE))
  oneminusprec <- sapply(resTest, function(z) {
      idx <- which(z$padj < .1)
      ifelse(sum(idx) == 0, 0, mean((beta == 0)[idx]))
  })
  data.frame(sensitivity=sens,
             sensStratified,
             oneminusspecpvals=oneminusspecpvals,
             oneminusspecpadj=oneminusspecpadj,
             oneminusprec=oneminusprec,
             algorithm=namesAlgos, effSize=es, m=m)
})
res <- do.call(rbind, resList)

ggplot(as_tibble(res,rownames="runNumber")) + geom_point(aes(x=oneminusprec,y=sensitivity,colour=algorithm)) + facet_grid(effSize~m)

#save(res, namesAlgos, file="../../data/simulated/results_simulateDE_comparison.RData")
```
# My model's performance
Redo the same analysis but with my model and add it to DESeq2 graph!
```{r,warning=FALSE}
register(MulticoreParam(workers=32)) 

source("~/diffExpModel/PhDBayesianWork/code/negativeBinomialModel/nbinDataandLibrarySetup.R")

algos <- list("DESeq2"=runDESeq2,"Bayesian"=openBayes,"voom"=runVoom)

cleanData <- cleanWildTypeData %>%
  inner_join(cleanMutantData,suffix = c(".W",".M"),by = "transcriptName")

WTData <- cleanData %>%
  select(ends_with(".W"))
MTData <- cleanData %>%
  select(ends_with(".M"))

namesAlgos <- names(algos)

mLevels <- c(4,8,16)
n <- 5755
nreps <- 4
ms <- rep(rep(mLevels, each=nreps),3)

# Read list of random sampled replicates used in Bayesian Analysis
bayesianReplicateList <- read_tsv("~/diffExpModel/PhDBayesianWork/data/simulated/fourRepdownSamplednBinCompMetaData.tsv")

# function to sample replicates from Schurch et al dataset
sampleExperimentalData <- function(m,k,n) {
  WTCols <- bayesianReplicateList %>%
    filter(str_detect(file,paste0("Sample",m/2,"Replicates",(k - 1)%%4 + 1))) %>% 
    pull(wildRep)
  MTCols <- bayesianReplicateList %>%
    filter(str_detect(file,paste0("Sample",m/2,"Replicates",(k - 1)%%4 + 1)))%>% 
    pull(mutRep)
  WTmatrix = as.matrix(WTData[1:n,WTCols])
  MTmatrix = as.matrix(MTData[1:n,MTCols])
  colnames(WTmatrix) <- NULL
  colnames(MTmatrix) <- NULL
  cbind(WTmatrix,MTmatrix)
}

# use 48 rep version of bayesian model as "ground truth"
groundBayesTruth = openBayes(ExpressionSet(matrix(nrow=n,ncol=32),annotation=as.character(1)))

# use 48 rep version of DESeq2 as "ground truth"
groundDEseq2Truth = runDESeq2(ExpressionSet(sampleExperimentalData(32,1,n),AnnotatedDataFrame(data.frame(condition = factor(rep(c("A","B"), each = 16))))))

# use 48 rep version of voom as "ground truth"
groundVoomTruth = runVoom(ExpressionSet(sampleExperimentalData(32,1,n),AnnotatedDataFrame(data.frame(condition = factor(rep(c("A","B"), each = 16))))))

betaBayes <- ( groundBayesTruth$beta > 0.6 & groundBayesTruth$padj < 0.00001)

betaDEseq2 <- (groundDEseq2Truth$beta > 0.1 & groundDEseq2Truth$padj < 0.00001)

betaVoom <- (groundVoomTruth$beta > 0.1 & groundVoomTruth$padj < 0.00001)

experimentalDataResList <- bplapply(seq_along(ms), function(i) {
  if(i < 13) {
    beta = betaDEseq2
    es = "DEseq2"
    }
  else if(i > 24) {
    beta = betaBayes
    es = "Bayes"
    }
  else {
    beta = betaVoom
    es = "Voom"
  }
  
  set.seed(i)
  m <- ms[i]
  condition <- factor(rep(c("A","B"), each = m/2))
  x <- model.matrix(~ condition)
  mat <- sampleExperimentalData(m,i,n)
  e <- ExpressionSet(mat, AnnotatedDataFrame(data.frame(condition)),annotation=as.character(i))
  resTest <- lapply(algos, function(f) f(e))
  nonzero <- rowSums(exprs(e)) > 0
  
  sensidx <- beta & nonzero
  sens <- sapply(resTest, function(z) mean((z$padj < .1 & z$beta > 0.1)[sensidx]))
  rmf <- cut(rowMeans(mat), c(0, 20, 100, 300, Inf), include.lowest=TRUE)
  levels(rmf) <- paste0("sens",c("0to20","20to100","100to300","more300"))
  sensStratified <- t(sapply(resTest, function(z) tapply( (z$padj < .1 & z$beta > 0.1)[sensidx], rmf[sensidx], mean)))
  oneminusspecpvals <- sapply(resTest, function(z) mean((z$pvals < .01 & z$beta > 0.1)[beta == 0 & nonzero], na.rm=TRUE))
  oneminusspecpadj <- sapply(resTest, function(z) mean((z$padj < .1 & z$beta > 0.1)[beta == 0 & nonzero], na.rm=TRUE))
  oneminusprec <- sapply(resTest, function(z) {
      idx <- which(z$padj < .1 & z$beta > 0.1)
      ifelse(sum(idx) == 0, 0, mean((beta == 0)[idx]))
  })
  data.frame(sensitivity=sens,
             sensStratified,
             oneminusspecpvals=oneminusspecpvals,
             oneminusspecpadj=oneminusspecpadj,
             oneminusprec=oneminusprec,
             algorithm=namesAlgos, effSize=es, m=m)
})

experimantalDataRes <- do.call(rbind, experimentalDataResList)

ggplot(as_tibble(experimantalDataRes,rownames="runNumber")) + geom_point(aes(x=oneminusprec,y=sensitivity,colour=algorithm)) + facet_grid(effSize~m)

```

